polyply/src/solvator.py

```diff
@@ -0,0 +1,138 @@
+            description
+        mol_idxs:
+            the indices of the solvent molecules to build
+        potential:
+            the form of the nonbonded potential
+        """
+        self.nonbond_matrix = nonbond_matrix
+        # concatenate all trees into one
+        self.nonbond_matrix.concatenate_trees()
+        # for convenience
+        self.box = self.nonbond_matrix.boxsize
+        self.max_force = max_force
+        self.mol_idxs = np.array(mol_idxs, dtype=int)
+        self.potential = potential
+
+    def clever_grid(self):
+        """
+        Given the largest sigma value of a collection of one bead per residue
+        description of solvents generate a grid that has as miniumum grid spacing
+        the distance at which the LJ force is at most the maximum force-criterion
+        provided. This also works for a mixed system of solvents and sizes, because
+        all cross interactions are geometric averages meaning they cannot exceed
+        the size of the self-interaction.
+        """
+        # find the largest self interaction sigma between two of the solvent beads
+        max_sigma = 0
+        for idx in self.mol_idxs:
+            gndx = self.nonbond_matrix.nodes_to_gndx[(idx, 0)]
+            atype_key = frozenset([self.nonbond_matrix.atypes[gndx]])
+            sigma = self.nonbond_matrix.interaction_matrix[atype_key][0]
+            if sigma > max_sigma:
+                max_sigma = sigma
+        # compute the minimum distance these beads need to have in order
+        # to fullfill the max-force criterion
+        min_dist = scipy.optimize.fsolve(norm_lennard_jones_force,
+                                         x0=0.2,
+                                         args=(max_sigma, 1.0, self.max_force))
+        print(min_dist)
+        # build grid
+        grid = np.mgrid[0:self.box[0]:min_dist,
+                        0:self.box[1]:min_dist,
+                        0:self.box[2]:min_dist].reshape(3, -1).T
+        return grid
+
+    def run_system(self, molecules):
+        """
+        Add coordinates for solvent molecules to a system as
+        defined by the molecule indices (self.mol_idxs).
+        """
+        # the solvent molecules to be placed
+        not_placed_sols = np.full((self.mol_idxs.shape[0]), True)
+        sol_idxs = np.arange(0, self.mol_idxs.shape[0])
+        sol_coords = np.zeros((self.mol_idxs.shape[0], 3))
+        # initialize the grid
+        grid = self.clever_grid()
+        indices = np.arange(0, len(grid))
+        # a boolean mask recording which grid points are used
+        avail_indices = np.full((len(grid)), True)
+        # the progress bar
+        pbar = tqdm(total=not_placed_sols.shape[0])
+        while np.sum(not_placed_sols) > 0:
+            rng = default_rng()
+            selected_idx = rng.choice(indices[avail_indices],
+                                      size=not_placed_sols.sum(),
+                                      replace=False)
```

Comment:
This looks /really/ slow. Consider making a list of indices in grid, shuffling those, and simply placing one-by-one (if no overlap with the polymers). The list of indices will also not be cheap though...
Plan C: turn the grid into a KDtree. Select all points in grid that are at least distance X from polymer KDtrees. Happy placing

Reply:
Plan C is definetly not good, because making the KDtree with a grid that is larger than needed is slow and consumes a LOT of memory. 

What makes this bit slow? The RNG selectiom of indices? At the moment performance wise we're talking  10-15min for 8Mio solvent particles and a virus.

Reply:
> Plan C is definetly not good, because making the KDtree with a grid that is larger than needed is slow and consumes a LOT of memory.

Ouch, that's also out then.

Doing `indices[avail_indices]` looks slow, since I expect this makes a copy in memory since the resulting array will not be contiguous in memory otherwise.
The downside of my plan B (list of indices) is that indices are integers, which consume more memory than an array of booleans. Use int32 indices, so you only end up using... 12 times more per index >< (3x32 for x and y and z, compared to 8 for a bool)
It's all horrible. There's definitely some dark magic to gain here though

Reply:
To be honest this solution is quite memory lean, I've been experimenting a bit with some formulations and using indexing with arrays works quite well. I think there is some numpy magic going on, where it creates a view or actually internally iterates over these things in a loop, or uses this SIMD stuff. No clue. 

Memory wise this part is not really a concern here. I profiled the memory by checking how much different stages increase memory usage. The max memory consumption is earlier in the code. For the 8Mio particle system the MetaMolecules and System end up occupying like 50% of the new machines memory. That's just loading them. The grid and these indices only add up to 5% more.

Reply:
Ok, leave it as-is then. If the program is too slow/memory intensive, let's optimize the slow/memory intensive parts.